src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2080,31 +2192,46 @@ public static double Max(this IEnumerable<double> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
             double value = 0;
-            bool hasValue = false;
-            foreach (double x in source)
+            
+            using (IEnumerator<double> e = source.GetEnumerator())
             {
-                if (hasValue)
-                {
-                    if (x > value || System.Double.IsNaN(value)) value = x;
-                }
-                else
+                if (!e.MoveNext()) throw Error.NoElements();
+                double x = e.Current;
+                if (double.IsNaN(x)) return double.NaN;
+                value = x;
+                
+                while (e.MoveNext())
                 {
-                    value = x;
-                    hasValue = true;
+                    x = e.Current;
+                    if (x > value) value = x;
+                    else if (Double.IsNaN(x)) return Double.NaN;
```

Comment:
@jasonwilliams200OK Since I reverted all my changes to `Min` and `Max`, the notation has gone back to `System.Single.NaN` and `System.Double.NaN`, from the original implementer.